### DISCUSSION TITLE
Resolving cyclic dependencies between modules in Langium

### DISCUSSION THREAD

**Discussion by mrtnstn** [eclipse-langium/langium#1021] [Q&A] [answered]
While working on a type system for my Langium language, I ran into this error message:
`Cycle detected. Please make "TypeSystem" lazy. See https://langium.org/docs/di/cyclic-dependencies`

It makes some sense to me, because I have both a TypeSystem and a TypeProvider module that depend on each other. 

My question is how to resolve this? How do I make the module 'lazy', as suggested in the error message? If all else fails, I could consider to merge the two modules into one, but I would prefer to keep some separation of concerns.

The referenced documentation does not seem to be written yet :-(

**Comment by msujew** [comment, marked as answer]
Hey @mrtnstn,

sometimes cycles are unavoidable (we have some [in our own framework](https://github.com/langium/langium/blob/ecdd03ba4af8f2fe284dc53454ab54f9671f1333/packages/langium/src/workspace/index-manager.ts#L92)). In order to make them lazy, you can do something like this:

```ts
class TypeSystem {
  private readonly provider: () => TypeProvider;
  constructor(services: TypeServices) {
    provider = () => services.TypeProvider; // <-- lazy evaluated service
  }
}

class TypeProvider {
  private readonly typeSystem: TypeSystem;
  constructor(services: TypeServices) {
    typeSystem = services.TypeSystem;
  }
}
```

**Reply by mrtnstn** [reply]
Thank you Mark for your speedy answer. That does the trick indeed. I'll see if I can find some time to write up that missing documentation.

**Comment by spoenemann** [comment]
There's an issue about writing the missing documentation page: https://github.com/langium/langium-website/issues/39
